File: src/main/java/com/minimalecommerce/ordering/application/CartService.java
package com.minimalecommerce.ordering.application;

import com.minimalecommerce.catalog.application.CatalogStockPort;
import com.minimalecommerce.catalog.domain.Product;
import com.minimalecommerce.identity.application.UserService;
import com.minimalecommerce.ordering.api.dto.AddCartItemRequest;
import com.minimalecommerce.ordering.api.dto.CartItemResponse;
import com.minimalecommerce.ordering.api.dto.CartResponse;
import com.minimalecommerce.ordering.domain.CartItem;
import com.minimalecommerce.ordering.infrastructure.CartItemRepository;
import com.minimalecommerce.shared.domain.BusinessException;
import com.minimalecommerce.shared.domain.NotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Service
public class CartService {

    private final CartItemRepository cartItems;
    private final CatalogStockPort catalog;
    private final UserService users;

    public CartService(CartItemRepository cartItems, CatalogStockPort catalog, UserService users) {
        this.cartItems = cartItems;
        this.catalog = catalog;
        this.users = users;
    }

    @Transactional(readOnly = true)
    public CartResponse get(UUID userId) {
        List<CartItemResponse> items = cartItems.findByUserIdOrderByAddedAtAsc(userId)
                .stream().map(CartItemResponse::from).toList();
        BigDecimal subtotal = items.stream()
                .map(CartItemResponse::subtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        return new CartResponse(items, items.size(), subtotal);
    }

    @Transactional
    public CartResponse add(UUID userId, AddCartItemRequest request) {
        Product product = catalog.requireActive(request.productId());
        if (product.getStock() < request.quantity()) {
            throw new BusinessException("STOCK_INSUFFICIENT", "No hay stock suficiente");
        }
        CartItem item = cartItems.findByUserIdAndProductId(userId, product.getId())
                .orElseGet(() -> {
                    CartItem created = new CartItem();
                    created.setUser(users.require(userId));
                    created.setProduct(product);
                    created.setUnitPrice(product.getPrice());
                    created.setQuantity(0);
                    return created;
                });
        int nextQty = item.getQuantity() + request.quantity();
        if (product.getStock() < nextQty) {
            throw new BusinessException("STOCK_INSUFFICIENT", "No hay stock suficiente");
        }
        item.setQuantity(nextQty);
        item.setUnitPrice(product.getPrice());
        cartItems.save(item);
        return get(userId);
    }

    @Transactional
    public CartResponse updateQuantity(UUID userId, UUID itemId, int quantity) {
        CartItem item = cartItems.findByIdAndUserId(itemId, userId)
                .orElseThrow(() -> new NotFoundException("ítem de carrito", itemId));
        if (item.getProduct().getStock() < quantity) {
            throw new BusinessException("STOCK_INSUFFICIENT", "No hay stock suficiente");
        }
        item.setQuantity(quantity);
        return get(userId);
    }

    @Transactional
    public CartResponse remove(UUID userId, UUID itemId) {
        CartItem item = cartItems.findByIdAndUserId(itemId, userId)
                .orElseThrow(() -> new NotFoundException("ítem de carrito", itemId));
        cartItems.delete(item);
        return get(userId);
    }

    @Transactional
    public void clear(UUID userId) {
        cartItems.deleteByUserId(userId);
    }
}

File: apps/api/src/modules/cart/cart.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../../shared/prisma/prisma.service';
import { InventoryService } from '../inventory/inventory.service';
import { AddCartItemDto } from './dto/cart.dto';
import { ApiException, NotFoundError } from '../../shared/errors/api-error';

@Injectable()
export class CartService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly inventory: InventoryService,
  ) {}

  async get(userId: string) {
    const items = await this.prisma.cartItem.findMany({
      where: { userId },
      include: { product: true },
      orderBy: { addedAt: 'asc' },
    });
    const mapped = items.map((i) => ({
      id: i.id,
      productId: i.productId,
      productName: i.product.name,
      quantity: i.quantity,
      unitPrice: Number(i.unitPrice),
      subtotal: Number(i.unitPrice) * i.quantity,
      availableStock: i.product.stock,
    }));
    const subtotal = mapped.reduce((a, b) => a + b.subtotal, 0);
    return { items: mapped, itemCount: mapped.length, subtotal };
  }

  async add(userId: string, dto: AddCartItemDto) {
    const product = await this.inventory.requireActive(dto.productId);
    if (product.stock < dto.quantity) {
      throw new ApiException('STOCK_INSUFFICIENT', 'No hay stock suficiente');
    }
    const existing = await this.prisma.cartItem.findUnique({
      where: { userId_productId: { userId, productId: product.id } },
    });
    const nextQty = (existing?.quantity ?? 0) + dto.quantity;
    if (product.stock < nextQty) {
      throw new ApiException('STOCK_INSUFFICIENT', 'No hay stock suficiente');
    }
    await this.prisma.cartItem.upsert({
      where: { userId_productId: { userId, productId: product.id } },
      create: {
        userId,
        productId: product.id,
        quantity: dto.quantity,
        unitPrice: product.price,
      },
      update: {
        quantity: nextQty,
        unitPrice: product.price,
      },
    });
    return this.get(userId);
  }

  async updateQuantity(userId: string, itemId: string, quantity: number) {
    const item = await this.prisma.cartItem.findFirst({
      where: { id: itemId, userId },
      include: { product: true },
    });
    if (!item) throw new NotFoundError('ítem de carrito', itemId);
    if (item.product.stock < quantity) {
      throw new ApiException('STOCK_INSUFFICIENT', 'No hay stock suficiente');
    }
    await this.prisma.cartItem.update({ where: { id: itemId }, data: { quantity } });
    return this.get(userId);
  }

  async remove(userId: string, itemId: string) {
    const item = await this.prisma.cartItem.findFirst({ where: { id: itemId, userId } });
    if (!item) throw new NotFoundError('ítem de carrito', itemId);
    await this.prisma.cartItem.delete({ where: { id: itemId } });
    return this.get(userId);
  }

  async clear(userId: string) {
    await this.prisma.cartItem.deleteMany({ where: { userId } });
  }
}
